src/Stratis.Bitcoin/P2P/PeerConnector.cs

```diff
@@ -0,0 +1,219 @@
+
+        private CancellationTokenSource disconnect;
+
+        /// <summary> How to calculate a group of an IP, by default using NBitcoin.IpExtensions.GetGroup.</summary>
+        private Func<IPEndPoint, byte[]> groupSelector;
+
+        internal int MaximumNodeConnections { get; set; }
+        private Network network;
+        internal NodeConnectionParameters ParentParameters { get; private set; }
+        internal RelatedNodeGroups RelatedGroups { get; set; }
+        internal NodeRequirement Requirements { get; private set; }
+
+        internal void AddNode(Node node)
+        {
+            Guard.NotNull(node, nameof(node));
+
+            this.ConnectedNodes.Add(node);
+        }
+
+        internal void ConnectToPeersAsync()
+        {
+            if (Interlocked.CompareExchange(ref this.busyConnectingToPeers, 1, 0) == 1)
+                return;
+
+            this.asyncLoop = this.asyncLoopFactory.Run(nameof(this.ConnectToPeersAsync), token =>
+            {
+                ConnectToPeers();
+                return Task.CompletedTask;
+            },
+            this.disconnect.Token,
+            repeatEvery: TimeSpans.RunOnce);
+        }
+
+        private void ConnectToPeers()
+        {
+            while (!this.disconnect.IsCancellationRequested)
+            {
+                if (this.ConnectedNodes.Count >= this.MaximumNodeConnections)
+                {
+                    Thread.Sleep(TimeSpan.FromSeconds(1));
+                    continue;
+                }
+
+                Node node = null;
+
+                try
+                {
+                    var peer = FindPeerToConnectTo();
+                    if (peer == null)
+                        continue;
+
+                    using (var timeoutTokenSource = CancellationTokenSource.CreateLinkedTokenSource(this.disconnect.Token))
+                    {
+                        timeoutTokenSource.CancelAfter(5000);
+
+                        this.ParentParameters.PeerAddressManager().PeerAttempted(peer.Endpoint, DateTimeOffset.Now);
+
+                        var clonedConnectParamaters = this.currentParameters.Clone();
+                        clonedConnectParamaters.ConnectCancellation = timeoutTokenSource.Token;
+
+                        node = Node.Connect(this.network, peer, clonedConnectParamaters);
+                        node.VersionHandshake(this.Requirements, timeoutTokenSource.Token);
+                    }
+                }
+                catch (Exception exception)
+                {
+                    if (node != null)
+                        node.DisconnectAsync("Error while connecting", exception);
+                }
+            }
+        }
+
+        private void Disconnect()
+        {
+            this.ConnectedNodes.DisconnectAll();
+        }
+
+        internal bool RemoveNode(Node node)
+        {
+            return this.ConnectedNodes.Remove(node);
+        }
+
+        private NetworkAddress FindPeerToConnectTo()
```

Comment:
I realise we did not consider the groups requirement to connect to specified peers.
I would suggest the following, we use inheritance to make the distinction.

`PeerConnector` by default will look at the AddressManager to find peers, we can then override `FindPeerToConnectTo` for `Connect` and `Addnode` 

We will add two new classes
- PeerConnectorAddnode : PeerConnector
- PeerConnectorConnect : PeerConnector

They will have an internal list of peers to connect to.

We can then also use the same technique for the `ByEndpoint` and `ByNetwork` delegates.

Thoughts?

Reply:
@dangershony I've had a think about this issue and I suspect it'll just be easier to filter the peer list in PeerAddressManager by how the peer was added.

PeerAddressManager already contains a lot of logic related to how peers are selected and adding this list to a PeerConnector class instance, will introduce more complexity. 

I propose we add an enum to PeerAddress called something like PeerIntroductionType and have the following types:

Discovered
Added
Exclusive

This tells us how the peer is "introduced" to PeerConnector to be connected to. Either through discovery, "-addnode" or an exclusive connection via "-connect".

Then when we ask PeerAddressManager to find a random to connect to we filter on based which PeerConnector was created (the enum will be injected into the class).

Thoughts?